### PR TITLE
feat: Simulator Base Project

### LINKED ISSUE #106: Set-up foundation for the simulator

### PR DESCRIPTION
**Pull request by AlfredoG87**
**Description**:
- Adds a new subproject for simulator
- Adds Configuration Framework.
- Integrates with Hedera-Protobuf and PBJ Types
- Implements DI Dagger across app
- Runs Entry point.
- JUnit
- Unit Tests for all above

**Related issue(s)**:

Fixes #106 

**Notes for reviewer**:

The Generator section is there as a placeholder intended to change on next PR, don't pay attention to either interface, impl or test related to Generator.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: simulator/src/main/java/com/hedera/block/simulator/config/types/GenerationMode.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package com.hedera.block.simulator.config.types;
+
+/** The GenerationMode enum defines the modes of generation for the block stream. */
+public enum GenerationMode {
+    /** Reads Blocks from a Folder. */
+    FOLDER,
+    /** Generates Blocks from rules */
+    SELF
```
Comment (mattp-swirldslabs) [inline simulator/src/main/java/com/hedera/block/simulator/config/types/GenerationMode.java:24]: What do you think about changing `SELF` to `AD_HOC` or `ADHOC`?

File: simulator/src/main/java/com/hedera/block/simulator/generator/MockBlockStreamManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ */
+
+package com.hedera.block.simulator.generator;
+
+import com.hedera.hapi.block.stream.Block;
+import javax.inject.Inject;
+
+/** A mock implementation of the BlockStreamManager interface. */
+public class MockBlockStreamManagerImpl implements BlockStreamManager {
```
Comment (mattp-swirldslabs) [inline simulator/src/main/java/com/hedera/block/simulator/generator/MockBlockStreamManagerImpl.java:23]: Why use the `Mock` prefix? Can this just be `BlockStreamManagerImpl`?

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import com.hedera.block.simulator.generator.BlockStreamManager;
+import com.swirlds.config.api.Configuration;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import javax.inject.Inject;
+
+/** BlockStream Simulator App */
+public class BlockStreamSimulatorApp {
+
+    private static final System.Logger LOGGER =
+            System.getLogger(BlockStreamSimulatorApp.class.getName());
```
Comment (mattp-swirldslabs) [inline simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:28]: I think you can change this LOGGER declaration to: `private final Logger LOGGER = System.getLogger(getClass().getName());`

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     * @return true if the block stream simulator is running, false otherwise
+     */
+    public boolean isRunning() {
+        return isRunning;
+    }
+
+    /** Stops the block stream simulator. */
+    public void stop() {
+        isRunning = false;
+        LOGGER.log(System.Logger.Level.INFO, "Block Stream Simulator has stopped");
```
Comment (mattp-swirldslabs) [inline simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:69]: You can use static imports to reduce `System.Logger.Level.INFO` to just `INFO`

**Review by mattp-swirldslabs — DISMISSED** [review]
You've got Dagger integrated, unit tests, etc. Looks good!

File: simulator/src/main/java/com/hedera/block/simulator/config/types/GenerationMode.java (hunk 1 above)
Comment (mattp-swirldslabs) [reply simulator/src/main/java/com/hedera/block/simulator/config/types/GenerationMode.java:24]: What about changing `FOLDER` to `DIR` to align with linux, etc?

Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/config/types/GenerationMode.java:24]: Again, also inherited those Enums but I like your suggestions, will change Folder for `DIR` and SELF for `ADHOC`

File: simulator/src/main/java/com/hedera/block/simulator/generator/MockBlockStreamManagerImpl.java (hunk 2 above)
Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/generator/MockBlockStreamManagerImpl.java:23]: This is not part of the current scope but needed an impl of the BlockStreamManager to make sure that the wiring was working on Dagger and the App.

However that impl will be removed in the next PR, for now is just a placeholder.

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java (hunk 3 above)
Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:28]: will do, thanks for the suggestion.

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java (hunk 4 above)
Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:69]: Thanks, will do! 👍

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java (hunk 3 above)
Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:28]: It does not work for static context or objects. 😅

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java (hunk 4 above)
Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:69]: ✅ Done!

File: simulator/src/main/java/com/hedera/block/simulator/config/data/GrpcConfig.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * The GrpcConfig class defines the configuration data for the gRPC client.
+ *
+ * @param serverAddress the address of the gRPC server
+ * @param port the port of the gRPC server
+ */
+@ConfigData("grpc")
+public record GrpcConfig(
+        @ConfigProperty(defaultValue = "localhost") @Min(0) @Max(65535) String serverAddress,
```
[the remark this answers is not in the document]
Comment (AlfredoG87) [reply simulator/src/main/java/com/hedera/block/simulator/config/data/GrpcConfig.java]: ✅  Done!

File: simulator/src/main/java/com/hedera/block/simulator/generator/MockBlockStreamManagerImpl.java (hunk 2 above)
Comment (mattp-swirldslabs) [reply simulator/src/main/java/com/hedera/block/simulator/generator/MockBlockStreamManagerImpl.java:23]: Sounds good

File: simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java (hunk 3 above)
Comment (mattp-swirldslabs) [reply simulator/src/main/java/com/hedera/block/simulator/BlockStreamSimulatorApp.java:28]: Ah, ok...thanks for taking a look

**Review by mattp-swirldslabs — APPROVED** [review]

**Review by rbarkerSL — APPROVED** [review]
Approving change to codecov.yml

**Pull request merged**